### PR TITLE
Add filter builder classes

### LINKED ISSUE #251: [Feature Request] Add Filter Builder
It would be helpful to have a class-based `filter` in `MeiliSearchIndex.search` instead of just raw string, e.g.:
`QueryOperator` in MongoDB https://github.com/kfiross/flutter_mongodb_realm/blob/master/lib/database/query_operator.dart

I can create a PR for this if there is a consensus on the design

### PR DESCRIPTION
**Pull request by ahmednfwela**
# Pull Request

This is a proof of concept for implementing  filter builder pattern

## Related issue
Fixes #251

## What does this PR do?
The target of this PR is to be able to do this:
```dart
await index.search(
  'prince',
  filterExpression: Meili.eq(
    Meili.attribute('tag'),
    Meili.value("Tale"),
  ),
);
//simplified version:
await index.search(
  'prince',
  filterExpression: 'tag'.toMeiliAttribute().eq("Tale".toMeiliValue()),
);
```
to achieve this, I added some new classes in `src/filter_builder`

## PR checklist
Please check if your PR fulfills the following requirements:
- [ ] Does this PR fix an existing issue, or have you listed the changes applied in the PR description (and why they are needed)?
- [ ] Have you read the contributing guidelines?
- [ ] Have you made sure that the title is accurate and descriptive of the changes?

Thank you so much for contributing to Meilisearch!

### CODE REVIEW

**Comment by ahmednfwela** [comment]
I went ahead and completed the entire implementation, I didn't add tests though, until these changes can be reviewed

**Comment by ahmednfwela** [comment]
cc @brunoocasali

**Comment by brunoocasali** [comment]
Thanks a lot @ahmednfwela 🤘. I'll take a look on it soon!

File: test/search_test.dart [hunk 1]
```diff
@@ -149,9 +165,19 @@ void main() {
...
-            await index.search('', filter: 'book_id < 100 AND tag = Tale');
+        var result = await index.search(
+          '',
+          filter: 'book_id < 100 AND tag = Tale',
+        );
+        var resultExpr = await index.search(
+          '',
+          filterExpression:
+              'book_id'.toMeiliAttribute().lt(100.toMeiliValue()).and(
+                    'tag'.toMeiliAttribute().eq("Tale".toMeiliValue()),
```
[the remark this answers is not in the document]
Comment (ahmednfwela) [reply test/search_test.dart]: that's an old one, I did remove them to separate file

File: lib/src/filter_builder/facade.dart [hunk 2]
```diff
@@ -0,0 +1,158 @@
+import 'filter_builder_base.dart';
+import 'operators.dart';
+import 'values.dart';
+
+class Meili {
+  static AttributeFilterExpression attribute(String path) =>
```
[the remark this answers is not in the document]
Comment (ahmednfwela) [reply lib/src/filter_builder/facade.dart:6]: I think it should be fine to shorten, similar to `equals` -> `eq` and `greaterThanEquals` -> `gte`

**Comment by ahmednfwela** [comment]
I pushed a new patch of commits, and renamed some classes to avoid conflicts and confusion

File: lib/src/filter_builder/operators.dart [hunk 3]
```diff
@@ -0,0 +1,236 @@
...
+  final List<MeiliValueExpressionBase> values;
+
+  const MeiliInOperatorExpression({
+    required this.attribute,
+    required this.values,
+  });
+
+  @override
+  String transform() {
+    //TODO(ahmednfwela): escape commas in values ?
```
Comment (brunoocasali) [inline lib/src/filter_builder/operators.dart:148]: Yeah, good catch!

File: pubspec.yaml [hunk 4]
```diff
@@ -11,6 +11,7 @@ environment:
 dependencies:
   dio: ^5.0.0
   crypto: ^3.0.1
+  collection: ^1.17.0
```
[the remark this answers is not in the document]
Comment (ahmednfwela) [reply pubspec.yaml:14]: as a general rule of thumb, dart doesn't compile what it doesn't need to execute, this is due to tree shaking, which will simply remove any unusable code, so adding a dependency is equivalent to copying and pasting the needed classes and functions directly into the library

File: lib/src/filter_builder/values.dart [hunk 5]
```diff
@@ -0,0 +1,93 @@
...
+    return value.replaceAll(r'\', r'\\').replaceAll(r"'", r"\'");
+  }
+
+  @override
+  String transform() {
+    final sb = StringBuffer();
+    sb.write("'");
+    sb.write(escapeValue(value));
+    sb.write("'");
+    return sb.toString();
```
[the remark this answers is not in the document]
Comment (ahmednfwela) [reply lib/src/filter_builder/values.dart:81]: that's interesting, I will look into it!

File: test/search_test.dart [hunk 6]
```diff
@@ -127,6 +128,19 @@ void main() {
...
       });
+      
+      test('filter parameter is null', () async {
+        var index = await createBooksIndex();
+        var response = await index
+            .updateSettings(IndexSettings(
+              filterableAttributes: ['tag'],
+            ))
+            .waitFor(client: client);
+        expect(response.status, 'succeeded');
```
Comment (brunoocasali) [inline test/search_test.dart:139]: Since you're using the `waitFor`, you'll receive a success status or the test it will fail anyway
```suggestion

```

**Comment by brunoocasali** [comment]
@ahmednfwela, I saw your comment in the engine's side PR regarding the `NULL` handling, and that's amazing to see how involved you are! 🤘 

But, if we keep that change, we must wait until Meilisearch v1.2 is released! Because until then, our CI will be broken, and I can't accept that on purpose.

So I think you should move that change to a new PR, and then we can merge this one safely. When Meilisearch v1.2 is out, I can merge the NULL improvement. WDYT?
Or, you can "skip" the test case and we can merge anyway, but I will have to deal with potential unsatisfied users who tried to used a feature that is not ready yet 😬

**Comment by ahmednfwela** [comment]
glad to be of use! I really like meilisearch and I will try to help as much as I can in any area possible.

also yes I agree on waiting until the NULL operator is released to add the feature back, so I have commented it for now

**Review by brunoocasali — APPROVED** [review]
@ahmednfwela, it is good to go from my POV!

I would like you to add a small sentence in the PR description or comment, so I can use it to add it to the CHANGELOG.

Thanks again, for your time with us!

**Comment by brunoocasali** [comment]
bors merge

**Comment by ahmednfwela** [comment]
I added some comments to summarize the changes

**Pull request merged**